[GStreamer][WebRTC] Rework replaceTrack handling

philn opened the pull request:
#### 88551187c932d1e173f270ba4df6e3f031d81e99
<pre>
[GStreamer][WebRTC] Rework replaceTrack handling
<a href="https://bugs.webkit.org/show_bug.cgi?id=280120">https://bugs.webkit.org/show_bug.cgi?id=280120</a>

Reviewed by Xabier Rodriguez-Calvar.

Track replacement is now handled in the GStreamer mediastream source element, instead of doing
convoluted caps negotiation issues-prone pipeline manipulations in outgoing sources. We used to have
an audiotestsrc/videotestsrc combined with input-selector there, but that could lead to caps
negotiation issues, for instance when replacing an audio track with a new one with different
capabilities (samplerate for instance).

* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpSenderBackend.cpp:
(WebCore::GStreamerRtpSenderBackend::replaceTrack):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
(mediaStreamTrackPrivateGetTags):
(webkitMediaStreamNew):
(WebKitMediaStreamObserver::didRemoveTrack):
(webkitMediaStreamSrcAddTrack):
(webkitMediaStreamSrcReplaceTrack):
(webkitMediaStreamSrcConfigureAudioTracks):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingAudioSourceGStreamer::setPayloadType):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::connectFallbackSource): Deleted.
(WebCore::RealtimeOutgoingAudioSourceGStreamer::unlinkOutgoingSource): Deleted.
(WebCore::RealtimeOutgoingAudioSourceGStreamer::linkOutgoingSource): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::RealtimeOutgoingMediaSourceGStreamer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::start):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stop):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stopOutgoingSource):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::sourceMutedChanged):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::sourceEnabledChanged):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::initializeFromTrack):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::teardown):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::replaceTrack):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setSource): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.h:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::track const):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::source const): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::connectFallbackSource): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::unlinkOutgoingSource): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::linkOutgoingSource): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingVideoSourceGStreamer::setPayloadType):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::connectFallbackSource): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::unlinkOutgoingSource): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::linkOutgoingSource): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/284065@main">https://commits.webkit.org/284065@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d14a04678dfbd1720853db78af71f507b18cf70

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68404 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47796 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21063 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72471 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19549 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70521 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55592 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19365 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13020 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71471 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/43699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/59068 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35079 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/40367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16488 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17906 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/62324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16836 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74163 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12375 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16102 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62067 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12414 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59146 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/62088 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15163 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10011 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3625 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43597 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44671 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45865 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44413 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
```diff
@@ -300,19 +326,26 @@ class InternalSource final : public MediaStreamTrackPrivateObserver,
     void configureAudioTrack(float volume, bool isMuted, bool isPlaying)
     {
-        ASSERT(m_track.isAudio());
+        [[maybe_unused]] RefPtr track = m_track.get();
```
Why do you need the unused tag here? It is used in the first if, right?

philn replied:
Yes :)

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
```diff
@@ -399,17 +435,17 @@ class InternalSource final : public MediaStreamTrackPrivateObserver,
     }
 
-    void trackEnabledChanged(MediaStreamTrackPrivate&) final
+    void trackEnabledChanged(MediaStreamTrackPrivate& track) final
```
Could this be const?

philn replied:
Well that would then imply all observer implementations to change. Can be done in a follow-up PR.

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
```diff
@@ -277,7 +299,7 @@ class InternalSource final : public MediaStreamTrackPrivateObserver,
     }
 
-    const MediaStreamTrackPrivate& track() const { return m_track; }
+    RefPtr<MediaStreamTrackPrivate> track() const { return m_track.get(); }
```
`WARN_UNUSED`

philn committed:
[GStreamer][WebRTC] Rework replaceTrack handling
https://bugs.webkit.org/show_bug.cgi?id=280120

Reviewed by Xabier Rodriguez-Calvar.

Track replacement is now handled in the GStreamer mediastream source element, instead of doing
convoluted caps negotiation issues-prone pipeline manipulations in outgoing sources. We used to have
an audiotestsrc/videotestsrc combined with input-selector there, but that could lead to caps
negotiation issues, for instance when replacing an audio track with a new one with different
capabilities (samplerate for instance).

* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpSenderBackend.cpp:
(WebCore::GStreamerRtpSenderBackend::replaceTrack):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
(mediaStreamTrackPrivateGetTags):
(webkitMediaStreamNew):
(WebKitMediaStreamObserver::didRemoveTrack):
(webkitMediaStreamSrcAddTrack):
(webkitMediaStreamSrcReplaceTrack):
(webkitMediaStreamSrcConfigureAudioTracks):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingAudioSourceGStreamer::setPayloadType):
(WebCore::RealtimeOutgoingAudioSourceGStreamer::connectFallbackSource): Deleted.
(WebCore::RealtimeOutgoingAudioSourceGStreamer::unlinkOutgoingSource): Deleted.
(WebCore::RealtimeOutgoingAudioSourceGStreamer::linkOutgoingSource): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingAudioSourceGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::RealtimeOutgoingMediaSourceGStreamer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::start):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stop):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stopOutgoingSource):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::sourceMutedChanged):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::sourceEnabledChanged):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::initializeFromTrack):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::teardown):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::replaceTrack):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setSource): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.h:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::track const):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::source const): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::connectFallbackSource): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::unlinkOutgoingSource): Deleted.
(WebCore::RealtimeOutgoingMediaSourceGStreamer::linkOutgoingSource): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingVideoSourceGStreamer::setPayloadType):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::connectFallbackSource): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::unlinkOutgoingSource): Deleted.
(WebCore::RealtimeOutgoingVideoSourceGStreamer::linkOutgoingSource): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.h:

Canonical link: https://commits.webkit.org/284065@main
